### ISSUE TITLE
The output of BatchNormalization may contain Nan under certain parameters

### ISSUE THREAD

**Issue by shiningrain** [tensorflow/tensorflow#38644]
**System information**  
- Have I written custom code (as opposed to using example directory):  
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):   Linux Ubuntu 18.04
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  2.1.0-CPU
- Python version:  3.6.9
- CUDA/cuDNN version:  -
- GPU model and memory:  -

**Describe the current behavior**  
I found that the output of  `BatchNormalization` in `tensorflow.keras` may contain "Nan" under certain parameters. As an operation for normalization calculations, this is quite abnormal for `BatchNormalization` with nan output. This may lead to horrible results for the models with `BatchNormalization`, for example, accidental termination in training.  Detailed reproduction can be found in the following codes.

![image](https://user-images.githubusercontent.com/46860123/79631824-f5a7ad00-818d-11ea-8474-c72228a6cf65.png)

## Code to reproduce the issue

```
import os
import numpy as np
import pickle
import argparse
import tensorflow as tf
import tensorflow.keras.layers as L
from tensorflow.keras.models import load_model


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='reproduce')
    parser.add_argument('-path', '-p', type=str,help='Path')
    args = parser.parse_args('-p ./Your Path/nan'.split( ))# Your path is where you unzip nan.zip
    root_path = os.path.abspath(args.path)
    with open(os.path.join(root_path, 'input.pkl'), 'rb') as f:#input,bug type,params
        meta = pickle.load(f)
    input1 = meta['input']
    input=input1.astype(np.float32)
    model_path = os.path.join(root_path, 'model.h5')
    model = load_model(model_path)
    output = model.predict(input)
    nanresult=np.isnan(output).any()
    print(nanresult)

```
[nan.zip](https://github.com/tensorflow/tensorflow/files/4494054/nan.zip)

Unzip the nan.zip, you can find the model.h5 and input.pkl in it.

**Comment by ravikyram** [comment]
@shiningrain 

I tried to reproduce the issue with TF version 2.1, 2.2-rc3. I am seeing this error message (`SystemExit: 2`). Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/b5532e13bc2eb5aa048eee145dada450/untitled790.ipynb). Is this the expected behavior?Thanks!

**Comment by shiningrain** [comment]
> @shiningrain
> 
> I tried to reproduce the issue with TF version 2.1, 2.2-rc3. I am seeing this error message (`SystemExit: 2`). Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/b5532e13bc2eb5aa048eee145dada450/untitled790.ipynb). Is this the expected behavior?Thanks!

Thanks for your reply!
As shown in the following figure, I this there is something wrong with your path for the unzipped nan.zip. It seem like colab can not find your path.
![image](https://user-images.githubusercontent.com/46860123/79747171-d3ed2800-833d-11ea-800c-0b4822a12f52.png)
I think the reason caused this problem may be the `split( )` in my code to split the string. Could you try to use a path without a  blank space? like `./content/drive/MyDrive/nan`, or may be you can use the following codes:
```
import os
import numpy as np
import pickle
import argparse
import tensorflow as tf
import tensorflow.keras.layers as L
from tensorflow.keras.models import load_model
tensorflowVersion=(tf.__version__)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='reproduce')
    parser.add_argument('-path', '-p', type=str,help='Path')
    args = parser.parse_args('-p$$/data/zxy/DL_framework/nan'.split("$$"))
    root_path = os.path.abspath(args.path)
    with open(os.path.join(root_path, 'input.pkl'), 'rb') as f:#input,bug type,params
        meta = pickle.load(f)
    input1 = meta['input']
    input=input1.astype(np.float32)
    model_path = os.path.join(root_path, 'model.h5')
    model = load_model(model_path)
    output = model.predict(input)
    nanresult=np.isnan(output).any()
    print(nanresult)
```

**Comment by ravikyram** [comment]
@shiningrain 

I tried to reproduce the issue with TF version 2.1.0. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/2602f1f2a94a08ccc6dc072198d19b18/untitled801.ipynb).Is this the expected behavior?Thanks!

**Comment by shiningrain** [comment]
> @shiningrain
> 
> I tried to reproduce the issue with TF version 2.1.0. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/2602f1f2a94a08ccc6dc072198d19b18/untitled801.ipynb).Is this the expected behavior?Thanks!

This is exactly the issue I want to show! Under such a model, `BatchNormalization` will return the results containing 'Nan', which will have horrible impact on actual use. I think that there should be some code defects in the implementation of `BatchNormalization`, because the calculation of `BatchNormalization` should be no possibility of overflow.
I hope you can confirm the root cause of this issue.
Thank you very much!

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.1.0, 2.2-rc3 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/63e5fdd38d1737e8b98142e0d4b9b984/untitled808.ipynb).Thanks!

**Comment by shiningrain** [comment]
> I have tried on colab with TF version 2.1.0, 2.2-rc3 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/63e5fdd38d1737e8b98142e0d4b9b984/untitled808.ipynb).Thanks!

Yes, that is it.
 I further find that when conducting `BatchNormalization`, Tensorflow will invoke the **sqrt** related operations. As shown in the following figure.
![image](https://user-images.githubusercontent.com/46860123/80238482-19d52380-8691-11ea-8856-34b0b9cc0b54.png)

 The **sqrt** returns NAN when taking negative values as inputs.  I conduct a confirmation testing with the following example. 
![image](https://user-images.githubusercontent.com/46860123/80238273-bba84080-8690-11ea-8c19-d4ae8922a7a8.png)

I think this should be the reason for the NAN output of `BatchNormalization` in the above model.
Tensorflow may need a check for such negative input when using sqrt to avoid possible nan output results that interfere with subsequent use.

**Comment by DNXie** [comment]
I tried this test case with 
`input = np.absolute(input)`, which makes sure all the values in the `input` are positive.
But it still gives `nan`. 
So there seems to be other causes.

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210525, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/aa9b10583ae04263001472f8fb173f9e/38644.ipynb). Thanks!

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
